src/communication/base_communicator.hpp

```diff
@@ -0,0 +1,149 @@
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <vector>
+#include <random>
+#include <functional>
+
+#include <algorithms.hpp>
+#include <connection.hpp>
+#include <communication/gathered_vector.hpp>
+#include <event_queue.hpp>
+#include <spike.hpp>
+#include <util/debug.hpp>
+#include <util/double_buffer.hpp>
+#include <util/partition.hpp>
+#include <util/compare.hpp>
+
+namespace nest {
+namespace mc {
+namespace communication {
+
+// When the communicator is constructed the number of target groups and targets
+// is specified, along with a mapping between local cell id and local
+// target id.
+//
+// The user can add connections to an existing communicator object, where
+// each connection is between any global cell and any local target.
+//
+// Once all connections have been specified, the construct() method can be used
+// to build the data structures required for efficient spike communication and
+// event generation.
+template <typename CommunicationPolicy>
+class base_communicator {
```

Comment:
I wanted to avoid defining the event_queue and the lt_spike 3 or 4 times as copy & paste, though.